widget/embedded/src/utils/colors.ts

```diff
@@ -1,141 +1,151 @@
+
+  return { red, green, blue };
+}
+
+// Generate a shade of a given RGB color
+function generateShade(rgb: RGB, step: number): RGB {
+  const factor = 1 - PERCENT * step;
+  return {
+    red: rgb.red * factor,
+    green: rgb.green * factor,
+    blue: rgb.blue * factor,
+  };
+}
+
+// Generate a tint of a given RGB color
+function generateTint(rgb: RGB, step: number): RGB {
+  const factor = PERCENT * step;
+  return {
+    red: rgb.red + (MAX_COLOR_VALUE - rgb.red) * factor,
+    green: rgb.green + (MAX_COLOR_VALUE - rgb.green) * factor,
+    blue: rgb.blue + (MAX_COLOR_VALUE - rgb.blue) * factor,
+  };
+}
 
-  return [r, g, b];
+// Main Functions
+
+// Calculate an array of shades from a hex color string
+export function calculateShades(hex: string): string[] {
+  return calculateColors(hex, generateShade);
 }
 
-// Function to clamp color component to valid range [0, 255] and convert to hex
-function clamp(color: number): string {
-  const MAX_COLOR_VALUE = 255;
-  const HEX_BASE = 16;
-  const HEX_LENGTH = -2;
-  return (
-    '0' + Math.max(0, Math.min(MAX_COLOR_VALUE, color)).toString(HEX_BASE)
-  )
-    .slice(HEX_LENGTH)
-    .toUpperCase();
+// Calculate an array of tints from a hex color string
+export function calculateTints(hex: string): string[] {
+  return calculateColors(hex, generateTint);
 }
 
-export function isBright(baseColor: string) {
-  const BRIGHTNESS_THRESHOLD = 188;
-  const RED_COEFFICIENT = 299;
-  const GREEN_COEFFICIENT = 587;
-  const BLUE_COEFFICIENT = 114;
-  const BRIGHTNESS_DIVISOR = 1000;
-
-  const [r, g, b] = convertToRGB(baseColor);
-  const brightness =
-    (r * RED_COEFFICIENT + g * GREEN_COEFFICIENT + b * BLUE_COEFFICIENT) /
-    BRIGHTNESS_DIVISOR;
-  if (brightness > BRIGHTNESS_THRESHOLD) {
-    return true;
+// General function to calculate shades or tints
+function calculateColors(
+  hex: string,
+  generateColor: ShadeOrTintFunction
+): string[] {
+  const rgb = hexToRGB(hex);
+  const colors: string[] = [];
+  for (let i = 1; i <= 9; i++) {
+    colors.push(rgbToHex(generateColor(rgb, i)));
   }
-  return false;
+  return colors;
 }
 
-export const generateColors = (
-  mainColors: { [x: string]: string },
+// Create tints and shades for a given color
+function createTintsAndShades(
+  hex: string,
+  prefix: string,
+  reverse: boolean
+): { [key: string]: string } {
+  const BASE_INDEX = 100;
+  const STEP = 50;
+
+  const tints = calculateTints(hex).reverse().concat(hex);
+  const shades = calculateShades(hex);
+  const combinedColors = tints.concat(shades);
+
+  const colorMap: { [key: string]: string } = {};
+  const length = combinedColors.length;
+
+  for (let i = 0; i < length; i++) {
+    const index = BASE_INDEX + (reverse ? length - 1 - i : i) * STEP;
+    colorMap[`${prefix}${index}`] = combinedColors[i];
+  }
+
+  return colorMap;
+}
+
+// Generate a color palette based on main colors and widget colors
+export function generateColors(
```

Comment:
Let's simplify this function and make it more clear:
1. Rename the function from `generateColros` to `extendThemeColors`.
2. Remove `shouldUpdate` logic and this function always should return colors.
3. As I said in a separate comment, `widgetColors` is mandatory for this function. Also rename the param to `extendedColors`.
4. Rename `mainColors` to `baseColors`.
4. Reorder params to be matched with this final function signature: `extendThemeColors(baseColors, extendedColors, options: { reverseNeutralRange  })`

Reply:
The final logic will only gets a base colors and will extend the base colors using second parameter (`extendedColors`) and always will return an object includes colors (so no empty object `{}`).